### PR TITLE
lib/model, lib/scanner: Prevent races aborting scans (fixes #6994)

### LINKED ISSUE #6994: "panic: filling Blocks: leveldb: snapshot released" on folder save
Following https://forum.syncthing.net/t/panic-filling-blocks-leveldb-snapshot-released/15638. This happened when I was trying to save changes to a folder using the Web GUI.

Syncthing version is `v1.10.0-rc.2` (x86-32). The OS is Windows 10 Enterprise 2016 LTSB x86. Not sure if of any significance, but the file system is NTFS, encrypted with BitLocker, and with NTFS compression enabled.

```
[7LDIB] 11:56:48 WARNING: Fatal error: ghgck-scplg Get(xxx): filling Blocks: leveldb: snapshot released
Panic at 2020-09-20T11:56:48+09:00
panic: filling Blocks: leveldb: snapshot released

goroutine 486054 [running]:
github.com/syncthing/syncthing/lib/db.fatalError(0x1599140, 0x13d425b0, 0x175f79e0, 0x5f, 0x1258e540)
	github.com/syncthing/syncthing/lib/db/set.go:535 +0x1e1
github.com/syncthing/syncthing/lib/db.(*FileSet).Snapshot.func1(0x1599140, 0x13d425b0, 0x175f79e0, 0x5f)
	github.com/syncthing/syncthing/lib/db/set.go:149 +0x43
github.com/syncthing/syncthing/lib/db.(*Snapshot).Get(0x15018d60, 0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff, 0x131fb6d0, ...)
	github.com/syncthing/syncthing/lib/db/set.go:241 +0x3ab
github.com/syncthing/syncthing/lib/model.cFiler.CurrentFile(0x15018d60, 0x131fb6d0, 0x4e, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/syncthing/syncthing/lib/model/folder.go:1153 +0x65
github.com/syncthing/syncthing/lib/scanner.(*walker).walkRegular(0x1411ff80, 0x15a2fc0, 0x15018da0, 0x131fb6d0, 0x4e, 0x15abae0, 0x1668a1c8, 0x14c28cc0, 0x0, 0x0)
	github.com/syncthing/syncthing/lib/scanner/walk.go:332 +0x5c
github.com/syncthing/syncthing/lib/scanner.(*walker).handleItem(0x1411ff80, 0x15a2fc0, 0x15018da0, 0x131fb6d0, 0x4e, 0x15abae0, 0x1668a1c8, 0x14c28cc0, 0x14c28d00, 0x0, ...)
	github.com/syncthing/syncthing/lib/scanner/walk.go:325 +0x1a3
github.com/syncthing/syncthing/lib/scanner.(*walker).walkAndHashFiles.func1(0x131fb6d0, 0x4e, 0x15abae0, 0x1668a1c8, 0x0, 0x0, 0x0, 0x0)
	github.com/syncthing/syncthing/lib/scanner/walk.go:268 +0xae2
github.com/syncthing/syncthing/lib/fs.(*MtimeFS).Walk.func1(0x131fb6d0, 0x4e, 0x15abae0, 0x1668a1c8, 0x0, 0x0, 0x8f198b, 0x1668a1c8)
	github.com/syncthing/syncthing/lib/fs/mtimefs.go:124 +0x71
github.com/syncthing/syncthing/lib/fs.(*walkFilesystem).walk(0x14958fa8, 0x131fb6d0, 0x4e, 0x15abae0, 0x1668a1c8, 0x155f63e0, 0x14155f50, 0x0, 0x0)
	github.com/syncthing/syncthing/lib/fs/walkfs.go:80 +0x13e
github.com/syncthing/syncthing/lib/fs.(*walkFilesystem).walk(0x14958fa8, 0x14b6e000, 0x43, 0x15abae0, 0x154cf548, 0x155f63e0, 0x14155f50, 0x0, 0x0)
	github.com/syncthing/syncthing/lib/fs/walkfs.go:112 +0x458
github.com/syncthing/syncthing/lib/fs.(*walkFilesystem).walk(0x14958fa8, 0x14c29040, 0x39, 0x15abae0, 0x154ce508, 0x155f63e0, 0x14155f50, 0x0, 0x0)
	github.com/syncthing/syncthing/lib/fs/walkfs.go:112 +0x458
github.com/syncthing/syncthing/lib/fs.(*walkFilesystem).walk(0x14958fa8, 0x12a36840, 0xd, 0x15abae0, 0x154ce410, 0x155f63e0, 0x14155f50, 0x0, 0x0)
	github.com/syncthing/syncthing/lib/fs/walkfs.go:112 +0x458
github.com/syncthing/syncthing/lib/fs.(*walkFilesystem).walk(0x14958fa8, 0x12d06f1, 0x1, 0x15abae0, 0x154ce2f8, 0x155f63e0, 0x14155f50, 0x0, 0x0)
	github.com/syncthing/syncthing/lib/fs/walkfs.go:112 +0x458
github.com/syncthing/syncthing/lib/fs.(*walkFilesystem).Walk(0x14958fa8, 0x12d06f1, 0x1, 0x155f63e0, 0x0, 0x155f63e0)
	github.com/syncthing/syncthing/lib/fs/walkfs.go:135 +0xe5
github.com/syncthing/syncthing/lib/fs.(*caseFilesystem).Walk(0x151692e0, 0x12d06f1, 0x1, 0x155f63e0, 0x34cd, 0x1ae2300)
	github.com/syncthing/syncthing/lib/fs/casefs.go:273 +0x88
github.com/syncthing/syncthing/lib/fs.(*MtimeFS).Walk(0x15018dc0, 0x12d06f1, 0x1, 0x1248e900, 0x14c28d00, 0x1248e900)
	github.com/syncthing/syncthing/lib/fs/mtimefs.go:110 +0x6c
github.com/syncthing/syncthing/lib/scanner.(*walker).walk.func1(0x1411ff80, 0x15a2fc0, 0x15018da0, 0x14c28cc0, 0x14c28d00)
	github.com/syncthing/syncthing/lib/scanner/walk.go:114 +0x1bf
created by github.com/syncthing/syncthing/lib/scanner.(*walker).walk
	github.com/syncthing/syncthing/lib/scanner/walk.go:111 +0x13e
```

[panic-20200920-115648.log](https://github.com/syncthing/syncthing/files/5251591/panic-20200920-115648.log)

### PR DESCRIPTION
**Pull request by imsodin**
This addresses two races which may lead to the panic reported in https://github.com/syncthing/syncthing/issues/6994:

One when a db error occurs: We immediately return in `folder.scanSubDirs`, not waiting for the scanner to terminate (close the returned chan).

Another when the scan is cancelled by context: The hasher aborts on context, and if it does so faster than the walker routine, it may close the returned channel while the walker is still working. Now the context is only checked when sending to the returned channel. Sends to "internal channels" are done without consideration of context, such that we can be sure everything is done when closing the returned channel.

I can't actually be sure this fixes #6994, as I can't deduct what exactly is happening from the trace, but the fixed scenarios might cause a panic with such a trace. Thus I believe it's fine to close that issue - we can always reopen if the same panic comes up again afterwards.

### CODE REVIEW

File: lib/scanner/blockqueue.go [hunk 1]
```diff
@@ -94,40 +94,35 @@ func (ph *parallelHasher) hashFiles(ctx context.Context) {
 	defer ph.wg.Done()
 
 	for {
-		select {
```
Comment (calmh) [inline lib/scanner/blockqueue.go:97]: Why remove the outer select? I guess the intent is that the inbox will be closed, but it looks odd that we only check the context on one "side" of the processing stage?

Comment (imsodin) [reply lib/scanner/blockqueue.go]: For one I somehow managed to commit the related changes in walk.go to another branch, making this change nonsensical anyway. My previous logic was to only check context on `outbox`/`finishedChan`, which is returned to the caller. That's unnecessary though, the only place we mustn't check context is the counter goroutine to not break the link between filesystem walking and hashing. Then wait on `ph.inbox` to be closed before closing `ph.outbox` to be sure filesystem walking finished when the returned channel gets closed.

File: lib/scanner/walk.go [hunk 2]
```diff
@@ -188,14 +188,13 @@ func (w *walker) walk(ctx context.Context) chan ScanResult {
 			}
 		}()
 
-	loop:
+		// We do not abort on context here because the link between the
+		// filesystem walker and the hasher must not be interrupted, otherwise
+		// the hasher might close finishedChan (returned to caller) before
+		// the filesystem walker is done.
```
Comment (calmh) [inline lib/scanner/walk.go:194]: I don't think that can happen? The walk is done as all files have been added to the `filesToHash` slice in the loop above on line 152-155. It shouldn't matter whether we send all of them or not

Comment (imsodin) [reply lib/scanner/walk.go]: Right - diff getting smaller again.

**Review by calmh — APPROVED** [review]

**Pull request merged**